apps/src/code-studio/progress.js

```diff
@@ -67,6 +68,25 @@ progress.showDisabledBubblesAlert = function () {
 };
 
 /**
+ * @returns {boolean}
+ */
+function getCachedIsUserSignedIn() {
+  // first check sessionStorage, where we may have cached sign in
+  let signedIn = clientState.getUserSignedIn();
+
+  // if that didn't give us anything, check the header, which uses a cookie-based
+  // approach to populate user id.
+  if (signedIn === null) {
+    const userData = $('.header_button.header_user.user_menu .user_name').data();
```

Comment:
I feel a _little_ funny about this.  What test can we put in place to keep this selector in sync with how this data is [described in user_header.haml](https://github.com/code-dot-org/code-dot-org/blob/78e526789e4d3f692e1dbd8bfa9dac3913e42e20/shared/haml/user_header.haml#L9)?  On a quick search I didn't see anything else that queries the user button for sign-in state.  Do we want this to be a more formally supported channel for passing data from Rails to JavaScript?

Less of an issue, but I always feel like calling it out - this is stuff that's easy to do without jQuery, using [`querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) and the [`dataset` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset).

Reply:
That's fair - I also don't feel great about this. One option would be to duplicate the code used to populate the DOM in `user_header.haml`. Sharing the code explicitly isn't easy given that it's embedded JS.

Another option might be to make the code in user_header.haml explicitly set some sessionStorage (that we then use here). I dont love adding more code to that though - especially since we can't easily test it.

I'm not really sure what the best approach is here.

Reply:
Sort of a weird idea, but I've thought a few times about building a `User` model in JavaScript that owned the knowledge about where to get this sort of information.  I might build a UI test that put us into a particular sign-in state, then sends a JavaScript command like `return new User().summarize()` to assert that the client-side model gets populated correctly.

Reply:
I'm not sure if you're using "I" here in the sense that you're considering doing this thing, or if you're suggesting that if you were in my shoes you might consider doing this thing.

For now, I think a change like you suggest is too big/out of scope for this PR. In the meantime, I've added comments to both locations to make it more clear that they need to be kept in sync.

Reply:
:+1: Sounds good.